### PR TITLE
docs: add `Basic Concepts` page

### PR DESCRIPTION
**Pull request by lzpap**
## Changes
- Separate `Basic Concepts` from `PyOTA Types`.
- Add a page describing basic IOTA concepts with references to the official docs site.

Internal links to PyOTA classes and methods point to parts of the documentation that will be updated in future commits.

### CODE REVIEW

File: docs/basic_concepts.rst [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+information is a ``trit``, that can have a value of -1, 0 or 1 in a balanced
+ternary system. A combination of 3 ``trits`` equals one ``tryte``, therefore a
+``tryte`` can have 3 * 3 * 3 = 27 different values.
+
+To represent a ``tryte``, IOTA encodes these 27 values into characters based on
+the `tryte alphabet`_.
+
+.. py:currentmodule:: iota
+
+In PyOTA, ``trits`` are represented as a sequence of numerical values (List[``int``])
```
Comment (todofixthis) [inline docs/basic_concepts.rst]:
```suggestion
In PyOTA, ``trits`` are represented as a sequence of numerical values (``List[int]``)
```

File: docs/basic_concepts.rst [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+`Seed`_ in IOTA is your unique password. It is the digital key that unlocks
+your safe that holds your tokens, or proves the ownership of messages.
+
+Seeds in PyOTA are always 81 trytes long and may only contain characters from
+the `tryte alphabet`_.
+
+.. warning::
+    Treat your seed(s) the same as you would the password for any other financial service.
+    Never share your seed with anyone and never use online generators to create a seed.
```
Comment (todofixthis) [inline docs/basic_concepts.rst]:
```suggestion
    Never share your seed with anyone, and never use online generators to create a seed.
```

File: docs/basic_concepts.rst [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    The library can help you to create your own locally and it does not require
+    internet connection: :py:meth:`iota.crypto.Seed.random`.
+
+For PyOTA specific implementation details on seeds, see :py:class:`crypto.Seed`.
```
Comment (todofixthis) [inline docs/basic_concepts.rst]:
```suggestion
For PyOTA-specific implementation details on seeds, see :py:class:`crypto.Seed`.
```

File: docs/basic_concepts.rst [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+on your entrance door: anyone can drop things in your mailbox (send you messages
+or tokens), but only you can empty it (withdraw tokens).
+
+.. warning::
+    Addresses should not be re-used once they are `spent from`_.
+    **You can receive as many transactions to an address as you wish, but only
+    spend from that address once.**
+
+    That being said, if you received money in your mailbox, you better move to a
+    new flat.
```
Comment (todofixthis) [inline docs/basic_concepts.rst]: ❔ I read this as using metaphor to point out how ridiculous the restriction is (possibly not the way it was intended), which may be counter-productive to the message of the warning.

File: docs/basic_concepts.rst [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+Addresses are generated from your seed through cryptographic functions. There
+are 9\ :sup:`57`\  different addresses that one might generate from a seed,
+which is quite a lot. Given your ``seed``, the ``index`` and a ``security level``
```
Comment (todofixthis) [inline docs/basic_concepts.rst]:
```suggestion
which is quite a lot. Given your ``seed``, the ``index`` and ``security level``
```

File: docs/basic_concepts.rst [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+parameters specify which address will be generated from it. The process is
+deterministic, meaning that same input paramteres always generate the same address.
+
+Addresses are 81 trytes long and may contain extra 9 trytes for checksum.
+The checksum may be used to verify that an address is in fact a valid IOTA address.
+
+For PyOTA specific implementation details on addresses, see :py:class:`Address`.
```
Comment (todofixthis) [inline docs/basic_concepts.rst]:
```suggestion
For PyOTA-specific implementation details on addresses, see :py:class:`Address`.
```

File: docs/basic_concepts.rst [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+Transaction
+-----------
+
+From the `official IOTA documentation site`_:
+
+*A transaction is a single transfer instruction that can either withdraw IOTA
+tokens from an address, deposit them into an address, or have zero-value
+(contain data, a message, or a signature). If you want to send anything to an
+IOTA network, you must send it to a node as a transaction.*
```
Comment (todofixthis) [inline docs/basic_concepts.rst]: Would a [`block-quote`](http://docutils.sourceforge.net/docs/ref/doctree.html#block-quote) be suitable here?

File: docs/basic_concepts.rst [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+in the transaction, the returning transaction hash would alter. This way, transaction
+hashes ensure the immutability of the Tangle.
+
+To become accepted by the network, a transaction has to be attached to the Tangle.
+The attachment process means that the transaction should reference two unconfirmed
+transactions (tips) in the Tangle and do a small proof-of-work. This process might
+be performed by a node, or by using the local proof-of-work feature of the `client
+libraries`_.
+
+For PyOTA specific implementation details on transactions, see :py:class:`Transaction`
```
Comment (todofixthis) [inline docs/basic_concepts.rst]:
```suggestion
For PyOTA-specific implementation details on transactions, see :py:class:`Transaction`
```

File: docs/basic_concepts.rst [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+.. note::
+    Unlike a block in a blockchain, bundles are not first-class
+    citizens in IOTA; only transactions get stored in the Tangle.
+
+    Instead, bundles must be inferred by following linked transactions
+    with the same bundle hash.
+
+Transactions in the bundle are linked together through their ``branchTransaction``
+and ``trunkTransaction`` fields, furthermore they are indexed within the bundle
```
Comment (todofixthis) [inline docs/basic_concepts.rst]: ❕ I believe the client libraries rely exclusively on `trunkTransaction` to link bundle transactions together; as far as I'm aware, `branchTransaction` does not have any function beyond confirming a second tip.

File: docs/basic_concepts.rst [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+Read more about `how bundles are structured`_.
+
+Bundles can be classified into two categories:
+
+- `Transfer bundles`_: Bundles that contain input and output transactions. A bundle
+  always has to be balanced, meaning that input transaction values should equal
+  to output transaction values.
+- `Zero-value bundles`_: Bundles that contain only zero-value transactions.
+
+For PyOTA specific implementation details on bundles, see :py:class:`Bundle`
```
Comment (todofixthis) [inline docs/basic_concepts.rst]:
```suggestion
For PyOTA-specific implementation details on bundles, see :py:class:`Bundle`
```

File: docs/basic_concepts.rst [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+and :py:class:`ProposedBundle`.
+
+Now that you are familiar with seom basic IOTA concepts, it is time to explore
```
Comment (todofixthis) [inline docs/basic_concepts.rst]:
```suggestion
Now that you are familiar with some basic IOTA concepts, it is time to explore
```

File: docs/basic_concepts.rst [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+.. _spent from: https://docs.iota.org/docs/getting-started/0.1/clients/addresses#spent-addresses
+.. _structure: https://docs.iota.org/docs/getting-started/0.1/transactions/transactions#structure-of-a-transaction
+.. _input transaction: https://docs.iota.org/docs/getting-started/0.1/transactions/transactions#input-transactions
+.. _output transaction: https://docs.iota.org/docs/getting-started/0.1/transactions/transactions#output-transactions
+.. _zero-value transaction: https://docs.iota.org/docs/getting-started/0.1/transactions/transactions#zero-value-transactions
+.. _client libraries: https://docs.iota.org/docs/client-libraries/0.1/introduction/overview
+.. _how bundles are structured: https://docs.iota.org/docs/getting-started/0.1/transactions/bundles
+.. _unique identifier for the bundle: https://docs.iota.org/docs/getting-started/0.1/transactions/bundles#bundle-hash
+.. _transfer bundles: https://docs.iota.org/docs/getting-started/0.1/transactions/bundles#transfer-bundles
+.. _zero-value bundles: https://docs.iota.org/docs/getting-started/0.1/transactions/bundles#zero-value-bundle
```
Comment (todofixthis) [inline docs/basic_concepts.rst]: ❔ Will we need to update these links every time a new version is published to `docs.iota.org`?

(e.g., `https://docs.iota.org/docs/getting-started/0.2`, `0.3`, etc.)

File: docs/types.rst [hunk 13]
```diff
@@ -1,11 +1,3 @@
-Basic Concepts
-==============
-
-Before diving into the API, it's important to understand the fundamental
-data types of IOTA.
-
-:todo: Link to IOTA docs
```
Comment (todofixthis) [inline docs/types.rst]: 😹 omg how long has that been there??

File: docs/basic_concepts.rst (hunk 9 above)
Comment (lzpap) [reply docs/basic_concepts.rst]: You are right. What do you think about including [such](https://docs.iota.org/assets/docs/getting-started/0.1/images/bundle-structure.png) a picture to illustrate this?

File: docs/basic_concepts.rst (hunk 12 above)
Comment (lzpap) [reply docs/basic_concepts.rst]: I'm afraid so. However, the docs site just has been updated to the current format after months of work, I think this getting started page will stay with us for a while.

File: docs/basic_concepts.rst (hunk 4 above)
Comment (lzpap) [reply docs/basic_concepts.rst]: Thanks for pointing this out! Gonna remove it then.

File: docs/basic_concepts.rst (hunk 7 above)
Comment (lzpap) [reply docs/basic_concepts.rst]: Yes, it looks nicer that way.

File: docs/basic_concepts.rst (hunk 9 above)
Comment (todofixthis) [reply docs/basic_concepts.rst]: Ooh, ya, that looks good!

File: docs/types.rst (hunk 13 above)
Comment (lzpap) [reply docs/types.rst]: little more than 2 years :smile_cat:

**Pull request merged**